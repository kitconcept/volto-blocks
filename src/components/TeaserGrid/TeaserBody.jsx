import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Message } from 'semantic-ui-react';
import { defineMessages, injectIntl } from 'react-intl';
import imageBlockSVG from '@plone/volto/components/manage/Blocks/Image/block-image.svg';
import { getContent } from '@plone/volto/actions';
import { flattenToAppURL } from '@plone/volto/helpers';

const messages = defineMessages({
  PleaseChooseContent: {
    id: 'Please choose an existing content as source for this element',
    defaultMessage:
      'Please choose an existing content as source for this element',
  },
});

const TeaserItem = ({ data, block, isEditMode, intl }) => {
  const contentSubrequests = useSelector(state => state.content.subrequests);
  const dispatch = useDispatch();
  const blockID = data.id || block;
  const result = contentSubrequests?.[blockID]?.data;

  React.useEffect(() => {
    if (data.href) {
      dispatch(getContent(data.href, null, blockID));
    }
  }, [dispatch, data, blockID]);

  return (
    <>
      {!data.href && (
        <Message>
          <div className="grid-teaser-item default">
            <img src={imageBlockSVG} alt="" />
            <p>{intl.formatMessage(messages.PleaseChooseContent)}</p>
          </div>
        </Message>
      )}
      {data.href && result && (
        <div className="grid-teaser-item top">
          {(() => {
            const item = (
              <>
                {result?.image && (
                  <img src={flattenToAppURL(result.image.download)} alt="" />
                )}
                <h3>{result.title}</h3>
                <p>{result.description}</p>
              </>
            );
            if (!isEditMode) {
              return (
                <Link
                  to={flattenToAppURL(result['@id'])}
                  target={data.openLinkInNewTab ? '_blank' : null}
                >
                  {item}
                </Link>
              );
            } else {
              return item;
            }
          })()}
        </div>
      )}
    </>
  );
};

TeaserItem.propTypes = {
  data: PropTypes.objectOf(PropTypes.any).isRequired,
  isEditMode: PropTypes.bool,
};

export default injectIntl(TeaserItem);
