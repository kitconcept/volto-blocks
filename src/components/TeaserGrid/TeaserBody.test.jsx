import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { MemoryRouter } from 'react-router-dom';

import TeaserBody from './TeaserBody';

const mockStore = configureStore();

test('renders a Grid Tile Proxy Item component', () => {
  const store = mockStore({
    content: {
      subrequests: {
        '2110a241-1389-4cda-8811-77031a540efa': {
          data: {
            title: 'the title',
            description: 'the description',
            image: { download: 'http://image' },
            '@id': 'http://theitem',
          },
        },
      },
    },
    intl: {
      locale: 'en',
      messages: {},
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <TeaserBody
          id="dcdf1f42-645d-48f6-9531-357bdc2e1881"
          data={{
            '@type': 'teaser',
            href: '/news/a-news-item-that-tells-things',
            id: '2110a241-1389-4cda-8811-77031a540efa',
            index: 0,
          }}
          tile="1234"
          onChangeTile={() => {}}
          openObjectBrowser={() => {}}
        />
      </MemoryRouter>
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
