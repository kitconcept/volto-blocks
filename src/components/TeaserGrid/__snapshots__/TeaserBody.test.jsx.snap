// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Grid Block Proxy Item component 1`] = `
<div
  className="grid-teaser-item top"
>
  <a
    href="/news/a-news-item-that-tells-things"
    onClick={[Function]}
    target={null}
  >
    <div
      className="grid-image-wrapper"
    >
      <img
        alt=""
        loading="lazy"
        src="/the-image.jpg"
      />
    </div>
    <h3>
      The title
    </h3>
    <p>
      The description
    </p>
  </a>
</div>
`;
